presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java

```diff
@@ -112,15 +114,35 @@ private boolean isMergeJoinEligible(PlanNode left, PlanNode right, JoinNode node
             // Acquire data properties for both left and right side
             StreamPropertyDerivations.StreamProperties leftProperties = StreamPropertyDerivations.derivePropertiesRecursively(left, metadata, session, types, parser);
             StreamPropertyDerivations.StreamProperties rightProperties = StreamPropertyDerivations.derivePropertiesRecursively(right, metadata, session, types, parser);
-
             List<VariableReferenceExpression> leftJoinColumns = node.getCriteria().stream().map(JoinNode.EquiJoinClause::getLeft).collect(toImmutableList());
-            List<VariableReferenceExpression> buildHashVariables = node.getCriteria().stream()
-                    .map(JoinNode.EquiJoinClause::getRight)
-                    .collect(toImmutableList());
+            List<VariableReferenceExpression> buildHashVariables = node.getCriteria().stream().map(JoinNode.EquiJoinClause::getRight).collect(toImmutableList());
+
+            // Check if the left side and right side's partitioning columns are subset of join columns
+            if (!verifyStreamProperties(leftProperties, leftJoinColumns) || !verifyStreamProperties(rightProperties, buildHashVariables)) {
+                return false;
+            }
 
             // Check if the left side and right side are both ordered by the join columns
-            return !LocalProperties.match(rightProperties.getLocalProperties(), LocalProperties.sorted(buildHashVariables, ASC_NULLS_FIRST)).get(0).isPresent() &&
-                    !LocalProperties.match(leftProperties.getLocalProperties(), LocalProperties.sorted(leftJoinColumns, ASC_NULLS_FIRST)).get(0).isPresent();
+            if (LocalProperties.match(rightProperties.getLocalProperties(), LocalProperties.sorted(buildHashVariables, ASC_NULLS_FIRST)).get(0).isPresent() ||
```

Comment:
There are cases where merge join can be planned but they cannot pass this check.
For example, if the two tables are both sorted by [A, B], and we join them on [B, A]. In this case, merge join should work. 
The order in the leftJoinColumns and buildHashVariables doesn't matter.
We may need to add another field preSortedColumns to the merge join node(as Masha suggested) to indicate which columns are pre-sorted. The order of these columns could be different from the join key orders. In the above example, the join keys are [B, A], the preSortedColumns are [A, B].
Adding this field will also make it possible to support partial merge joins.

This can be done in a separate PR as it is a new feature.

Reply:
I addressed this comment by adding new logic; Can you help review if the new logic makes sense?